spring-integration-core/src/main/java/org/springframework/integration/handler/DelayHandler.java

```diff
@@ -365,9 +367,16 @@ private void doReleaseMessage(Message<?> message) {
 
 	private boolean removeDelayedMessageFromMessageStore(Message<?> message) {
 		if (this.messageStore instanceof SimpleMessageStore) {
-			SimpleMessageGroup messageGroup =
-					(SimpleMessageGroup) this.messageStore.getMessageGroup(this.messageGroupId);
-			return messageGroup.remove(message);
+			synchronized (this.messageGroupId) {
+				Collection<Message<?>> messages = this.messageStore.getMessageGroup(this.messageGroupId).getMessages();
+				if (messages.contains(message)) {
```

Comment:
Looks like it was intentional to support concurrency....

https://github.com/spring-projects/spring-integration/commit/7bcadf6e9025d11b9d59ea22bdfd289e286920be

...poor decision.